drivers/led/sn32/rgb_matrix_sn32f24xb.c

```diff
@@ -0,0 +1,424 @@
+#include "rgb_matrix.h"
+#include "sn32f24xb.h"
+
+#if !defined(RGB_MATRIX_HUE_STEP)
+#    define RGB_MATRIX_HUE_STEP 8
+#endif
+
+#if !defined(RGB_MATRIX_SAT_STEP)
+#    define RGB_MATRIX_SAT_STEP 16
+#endif
+
+#if !defined(RGB_MATRIX_VAL_STEP)
+#    define RGB_MATRIX_VAL_STEP 16
+#endif
+
+#if !defined(RGB_MATRIX_SPD_STEP)
+#    define RGB_MATRIX_SPD_STEP 16
+#endif
+
+#ifdef MATRIX_NO_SCAN
+#    define ROWS_PER_HAND (MATRIX_ROWS)
+
+#    ifndef MATRIX_IO_DELAY
+#        define MATRIX_IO_DELAY 30
+#    endif
+#endif
+/*
+    COLS key / led
+    SS8050 transistors NPN driven low
+    base      - GPIO
+    collector - LED Col pins
+    emitter   - VDD
+
+    VDD     GPIO
+    (E)     (B)
+     |  PNP  |
+     |_______|
+         |
+         |
+        (C)
+        LED
+
+    ROWS RGB
+    SS8550 transistors PNP driven high
+    base      - GPIO
+    collector - LED RGB row pins
+    emitter   - GND
+
+        LED
+        (C)
+         |
+         |
+      _______
+     |  NPN  |
+     |       |
+    (B)     (E)
+    GPIO    GND
+*/
+static uint8_t chan_col_order[LED_MATRIX_COLS] = {0}; // track the channel col order
+static uint8_t current_row                     = 0;   // LED row scan counter
+static uint8_t row_idx                         = 0;   // key row scan counter
+#ifdef MATRIX_NO_SCAN
+extern matrix_row_t raw_matrix[MATRIX_ROWS];    // raw values
+extern matrix_row_t matrix[MATRIX_ROWS];        // debounced values
+static matrix_row_t shared_matrix[MATRIX_ROWS]; // scan values
+static bool         matrix_locked  = false;     // matrix update check
+static bool         matrix_scanned = false;
+#endif
+static const uint32_t periodticks                      = 256;
+static const uint32_t freq                             = (RGB_MATRIX_HUE_STEP * RGB_MATRIX_SAT_STEP * RGB_MATRIX_VAL_STEP * RGB_MATRIX_SPD_STEP * RGB_MATRIX_LED_PROCESS_LIMIT);
+static const pin_t    led_row_pins[LED_MATRIX_ROWS_HW] = LED_MATRIX_ROW_PINS; // We expect a R,B,G order here
+static const pin_t    led_col_pins[LED_MATRIX_COLS]    = LED_MATRIX_COL_PINS;
+static RGB            led_state[RGB_MATRIX_LED_COUNT]; // led state buffer
+bool                  enable_pwm = false;
+#ifdef UNDERGLOW_RBG // handle underglow with flipped B,G channels
+static const uint8_t underglow_leds[UNDERGLOW_LEDS] = UNDERGLOW_IDX;
+#endif
+
+#ifdef MATRIX_NO_SCAN
+void matrix_output_unselect_delay(uint8_t line, bool key_pressed) {
+    for (int i = 0; i < TIME_US2I(MATRIX_IO_DELAY); ++i) {
+        __asm__ volatile("" ::: "memory");
+    }
+}
+bool is_matrix_locked(void) {
+    return matrix_locked;
+}
```

Comment:
A more understandable and more correct API would be:
```c
bool matrix_available(void) {
    return matrix_scanned;
}
```
Using `matrix_locked` here might be racy — e.g., immediately after `matrix_scan_custom()` returns the state would be `matrix_scanned == false, matrix_locked == false`, so the `matrix_task()` code may incorrectly allow another `matrix_scan()` again if it looks at `matrix_locked` before the next PWM interrupt comes.

With the existing `is_matrix_locked()` API the correct way would probably be:
```suggestion
bool is_matrix_locked(void) {
    return !matrix_scanned;
}
```